pkg/router/template/router.go

```diff
@@ -430,23 +439,29 @@ func (r *templateRouter) writeDefaultCert() error {
 func (r *templateRouter) watchMutualTLSCert() error {
 	caPath := os.Getenv("ROUTER_MUTUAL_TLS_AUTH_CA")
 	if len(caPath) != 0 {
-		reloadFn := func() {
-			log.V(0).Info("reloading to get updated client CA", "name", caPath)
-			r.rateLimitedCommitFunction.RegisterChange()
+		r.haveClientCA = true
+		if err := crl.InitMTLSDirectory(caPath); err != nil {
+			return err
 		}
-		if err := r.watchVolumeMountDir(filepath.Dir(caPath), reloadFn); err != nil {
-			log.V(0).Info("failed to establish watch on mTLS certificate directory", "error", err)
-			return nil
+		haveCRLs, err := crl.CABundleHasCRLs(caPath)
+		if err != nil {
+			log.V(0).Error(err, "failed to parse CA Bundle", "path", caPath)
+			return err
 		}
-	}
-	crlPath := os.Getenv("ROUTER_MUTUAL_TLS_AUTH_CRL")
-	if len(crlPath) != 0 && filepath.Dir(caPath) != filepath.Dir(crlPath) {
-		reloadFn := func() {
-			log.V(0).Info("reloading to get updated client CA CRL", "name", crlPath)
+		r.haveCRLs = haveCRLs
+		caUpdateChannel := make(chan struct{})
+		crlReloadFn := func(haveCRLs bool) {
+			r.haveCRLs = haveCRLs
+			log.V(0).Info("reloading to get updated client CA CRL", "name", crl.CRLFilename, "have CRLs", haveCRLs)
 			r.rateLimitedCommitFunction.RegisterChange()
 		}
-		if err := r.watchVolumeMountDir(filepath.Dir(crlPath), reloadFn); err != nil {
-			log.V(0).Info("failed to establish watch on mTLS certificate directory", "error", err)
+		crl.ManageCRLs(caPath, caUpdateChannel, crlReloadFn)
+		caReloadFn := func() {
+			// Send signal to CRL management goroutine that client CA has been changed
+			caUpdateChannel <- struct{}{}
+		}
+		if err := r.watchVolumeMountDir(filepath.Dir(caPath), caReloadFn); err != nil {
+			log.V(0).Error(err, "failed to establish watch on mTLS certificate directory")
 			return nil
```

Comment:
Why does this not return the error?

Reply:
@Miciah I noticed that `writeDefaultCert` doesn't return an error for its watch on calls to `watchVolumeMountDir()` either - do you know of any history for why this returns nil?

Reply:
If we were to return an error I'd propose we do that change in 4.14+ only for the moment as we want the backports to discretely fix OCPBUGS-6661 and OCPBUGS-9464 only.